File: rs/minsk/src/analysis/syntax/node/expression/binary.rs
use crate::analysis::syntax::token::SyntaxToken;

use super::ExpressionSyntax;

#[derive(Debug)]
pub struct BinaryExpressionSyntax {
    pub left: Box<ExpressionSyntax>,
    pub operator_token: SyntaxToken,
    pub right: Box<ExpressionSyntax>,
}

File: py/minsk/analysis/syntax/expressions/binary.py
from collections.abc import Iterable
from dataclasses import dataclass

from minsk.analysis.syntax.expression import ExpressionSyntax
from minsk.analysis.syntax.kind import SyntaxKind
from minsk.analysis.syntax.node import SyntaxNode
from minsk.analysis.syntax.token import SyntaxToken


@dataclass(frozen=True)
class BinaryExpressionSyntax(ExpressionSyntax):
    left: ExpressionSyntax
    operator_token: SyntaxToken
    right: ExpressionSyntax

    @property
    def kind(self) -> SyntaxKind:
        return SyntaxKind.BinaryExpression

    @property
    def children(self) -> Iterable[SyntaxNode]:
        yield self.left
        yield self.operator_token
        yield self.right
